Firestore/core/src/firebase/firestore/remote/datastore.mm

```diff
@@ -99,16 +120,102 @@
 
 std::shared_ptr<WatchStream> Datastore::CreateWatchStream(
     id<FSTWatchStreamDelegate> delegate) {
-  return std::make_shared<WatchStream>(worker_queue_, credentials_, serializer_,
+  return std::make_shared<WatchStream>(worker_queue_, credentials_,
+                                       serializer_bridge_.GetSerializer(),
                                        &grpc_connection_, delegate);
 }
 
 std::shared_ptr<WriteStream> Datastore::CreateWriteStream(
     id<FSTWriteStreamDelegate> delegate) {
-  return std::make_shared<WriteStream>(worker_queue_, credentials_, serializer_,
+  return std::make_shared<WriteStream>(worker_queue_, credentials_,
+                                       serializer_bridge_.GetSerializer(),
                                        &grpc_connection_, delegate);
 }
 
+void Datastore::LookupDocuments(
```

Comment:
Any suggestions on how to make this cleaner are very welcome. This isn't horribly worse than [Objective-C version](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Source/Remote/FSTDatastore.mm#L256-L334). It's certainly significantly more verbose than Android and Web, but those platforms have two advantages:
* they use tasks/promises, which are nicer to chain than callbacks;
* in Android's case, authentication is handled by the gRPC client.

Reply:
Tried splitting into 3: `LookupDocuments` -> `LookupDocumentsWithToken` -> `OnLookupDocumentsResponse`, let me know what you think.

Also -- that boilerplate with `shared`/`weak_ptr` is required pretty much every time `GetToken` is called. Perhaps it could be moved into `CredentialsProvider` itself, or into a shared helper function?
```cpp
template <typename T>
void GetToken(std::weak_ptr<T>&& guard, AsyncQueue* queue, std::function<void(StatusOr<Token>)>&& on_token);
```